### PR TITLE
Update to Dart `1.20.0-dev.1.0`.

### PR DESCRIPTION
**Pull request by pq**
Update to Dart `1.20.0-dev.1.0` and corresponding `analyzer` package.
- updates misc. packages to make the version constraint solver happy
- makes a few source updates to keep up with new strong mode warnings.

Not ready to be integrated as is...  I started making some fixes to address new strong-mode warnings but really there are some choices that are best made by the maintainers.

/cc @Hixie @abarth 

Feel free to tag owners who may have an opinion about preferred fixes.

```
[warning] The element type '(ImmediateMultiDragGestureRecognizer) → ImmediateMultiDragGestureRecognizer' cannot be assigned to the map value type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (examples/flutter_gallery/lib/demo/animation_demo.dart, line 187, col 46)
[warning] The element type '(ImmediateMultiDragGestureRecognizer) → ImmediateMultiDragGestureRecognizer' cannot be assigned to the map value type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (examples/flutter_gallery/lib/demo/animation_demo.dart, line 353, col 46)
[warning] The element type 'PopupMenuDivider' cannot be assigned to the list type 'PopupMenuEntry<LeaveBehindDemoAction>' (examples/flutter_gallery/lib/demo/leave_behind_demo.dart, line 146, col 15)
[warning] The element type 'PopupMenuDivider' cannot be assigned to the list type 'PopupMenuEntry<String>' (examples/flutter_gallery/lib/demo/menu_demo.dart, line 143, col 17)
[warning] The return type 'SynchronousFuture<Map>' is not a 'Future<Map<String, List<String>>>', as defined by the method '_manifestParser'. (packages/flutter/lib/src/services/image_resolution.dart, line 147, col 12)
[warning] A value of type '(PanGestureRecognizer) → PanGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 284, col 40)
[warning] A value of type '(ScaleGestureRecognizer) → ScaleGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 295, col 42)
[warning] A value of type '(HorizontalDragGestureRecognizer) → HorizontalDragGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 269, col 51)
[warning] A value of type '(DoubleTapGestureRecognizer) → DoubleTapGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 236, col 46)
[warning] A value of type '(LongPressGestureRecognizer) → LongPressGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 243, col 46)
[warning] A value of type '(VerticalDragGestureRecognizer) → VerticalDragGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 254, col 49)
[warning] A value of type '(TapGestureRecognizer) → TapGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 226, col 40)
[warning] The element type '(HorizontalDragGestureRecognizer) → HorizontalDragGestureRecognizer' cannot be assigned to the map value type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/scrollable.dart, line 706, col 46)
[warning] The element type '(VerticalDragGestureRecognizer) → VerticalDragGestureRecognizer' cannot be assigned to the map value type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/scrollable.dart, line 696, col 44)
[warning] The return type 'State<StatefulWidget>' is not a 'T', as defined by the method 'currentState'. (packages/flutter/lib/src/widgets/unique_widget.dart, line 29, col 12)
```

### CODE REVIEW

**Comment by Hixie** [comment]
Most of those errors are errors in the analyzer, not the code. The fifth is the only one that I think might be legit, and even that one is a bit dubious IMHO.

File: packages/flutter_driver/lib/src/extension.dart [hunk 1]
```diff
@@ -179,23 +179,26 @@ class FlutterDriverExtension {
     return constructor(finder);
   }
 
-  Future<TapResult> tap(Tap command) async {
-    prober.tap(await _waitForElement(_createFinder(command.finder)));
+  Future<TapResult> tap(Command command) async {
+    Tap tap = command;
```
Comment (Hixie) [inline packages/flutter_driver/lib/src/extension.dart]: This makes the code worse. We shouldn't make this kind of change.

Comment (Hixie) [inline packages/flutter_driver/lib/src/extension.dart]: (Specifically, this removes the ability for the analyzer to tell when the method is being called incorrectly.)

Comment (pq) [inline packages/flutter_driver/lib/src/extension.dart]: Absolutely, I agree.  I should have prefaced this PR with the caveat that this is _"NOT how I'd like to do it but I don't see any other way to make strong-mode happy, please help!"_

Comment (pq) [inline packages/flutter_driver/lib/src/extension.dart]: I've reverted so you can see the errors in the proper context.

**Comment by Hixie** [comment]
@leafpetersen help! :-)

**Comment by pq** [comment]
I was just going to tag Leaf.  You beat me to it! 😄

**Comment by leafpetersen** [comment]
Taking a look.  What version are you updating from?

**Comment by pq** [comment]
Thanks!

`1.19.0-dev.5.0`

**Comment by leafpetersen** [comment]
Okay, I think all of these are places where strong mode errors have been integrated into the core analyzer.  We used to just issue generic STRONG_MODE_STATIC_TYPE_ERROR messages for a number of issues.  We've now integrated them into the core analyzer, so you now get the actual precise error.  So for example, for the first one, you now get a  MAP_VALUE_TYPE_NOT_ASSIGNABLE.

Unfortunately, you guys have been filtering based on that general STRONG_MODE_STATIC_TYPE_ERROR.

All of these except the 5th and the last error fall under the general category of issues that we identified in the Flutter Strong Mode doc as something that we'd like to have a way to support, but we don't converged on a language level solution yet.

For the callback cases, you can avoid the error by making the cast local:

```dart
     gestures: <Type, GestureRecognizerFactory>{
        ImmediateMultiDragGestureRecognizer: (GestureRecognizer _recognizer) {
         ImmediateMultiDragGestureRecognizer recognizer = _recognizer; 
          return (recognizer ??= new ImmediateMultiDragGestureRecognizer())
            ..onStart = _handleOnStart;
        }
      },
```
This is essentially the code that we would most likely end up having the compiler generate for you if we supported this pattern.

**Comment by abarth** [comment]
@leafpetersen How do you suggest we move forward here?

**Comment by leafpetersen** [comment]
I don't know what options are available to you in terms of just suppressing the errors - @pq will know better there.

In terms of fixing the errors:

The fifth and last errors I don't anticipate ever being strong mode compatible.  You can put a cast in there for now, but at least the fifth one will eventually start failing when the language moves to 2.0.  Using a `Map<dynamic, dynamic>` where a `Map<String, List<String>>` is expected breaks all of the assumptions that strong mode is trying to enforce to make AOT better.  

The second and third errors I could look into providing some experimental support for.  Not sure whether that fits your timeline for rolling, but I don't think it would take me too long to implement that.  Alternatively you can cast them, which will work for now.

A solution for the rest of the errors are probably a bit more work, and it's not clear that a solution that we put in now will be the final solution.  @jmesserly has started experimental work on supporting covariant overrides, and I have proposed to use a similar mechanism for these cases: basically you would annotate the parameter and the compiler would automatically generate the code that I wrote above.  It's possible that we could get something out the door for you guys to prototype in a week or two, but in the meantime you would either need to work around it using the pattern that I wrote above, or else by suppressing the individual errors in some way.

**Comment by abarth** [comment]
My understanding previously was that the Dart team was committed to working with us to evolving strong mode and the Flutter codebase to be compatible and that we'd made changes to each that were mutually agreeable.  With this change to the SDK, it seems like you're unilaterally forcing changes in the Flutter codebase (i.e., that are not mutually agreeable), contrary that understanding.

Perhaps a better path forward would be to revert whatever changes were made to the analyzer that have caused this regression and return to that agreed operating mode.

**Comment by leafpetersen** [comment]
We are very committed to working with the Flutter team to make the Dart language work well for their use cases.  We, and I personally, have devoted quite a bit of time this quarter towards making this happen, and I am personally very enthusiastic about having this kind of large scale feedback on the language early on.  A couple of things we should be clear on though, to be sure that we're on the same page.

1) This is not a strong mode breaking change. This is Flutter relying on an implementation detail of the analyzer error codes in order to suppress existing error messages.  Implementation details like that are not part of the public API, and can and will change.  I am happy to work *with* the Flutter team to try to accommodate their needs, but we cannot maintain anything close to the kind of velocity that we need to have if we have to run any change to our implementation by the Flutter team, and revert any change that touches on Flutter.

2) Strong mode has made, and will continue in the future to make, actual breaking changes as we move towards Dart 2.0.  Some of these breaking changes will affect Flutter, just as they currently affect large numbers of other users inside of google.  Sometimes those breaking changes will require changes to user code.  Sometimes the feedback we get from those breaking changes will lead us to back out of, or change our approach to, the breaking change.  None of this process is instantaneous, and if you are committed to rolling in new versions of strong mode frequently, you will certainly have to change code occasionally (hopefully usually for the better), and it is highly likely that you will sometimes need to work around issues.  The position that you should be able to roll strong mode weekly and never have to change your code is not tenable, and no commitment to such a thing has been made by, or communicated to, me.

**Comment by abarth** [comment]
> This is not a strong mode breaking change. This is Flutter relying on an implementation detail of the analyzer error codes in order to suppress existing error messages.

I'm not sure I understand.  Are you saying that the `strong_mode_static_type_error` field in `.analysis_options` is considered an implementation detail of the analyzer?  My understanding was that the `.analysis_options` file was an appropriate and supported way to configure the analyzer.  For example, the `.analysis_options` file is discussed extensively in https://pub.dartlang.org/packages/analyzer.

> we cannot maintain anything close to the kind of velocity that we need to have if we have to run any change to our implementation by the Flutter team, and revert any change that touches on Flutter.

That's not what I'm asking for.

> The position that you should be able to roll strong mode weekly and never have to change your code is not tenable, and no commitment to such a thing has been made by, or communicated to, me.

That's also not my position.

**Comment by leafpetersen** [comment]
> I'm not sure I understand. Are you saying that the strong_mode_static_type_error field in .analysis_options is considered an implementation detail of the analyzer? My understanding was that the .analysis_options file was an appropriate and supported way to configure the analyzer. For example, the .analysis_options file is discussed extensively in https://pub.dartlang.org/packages/analyzer.

The use of that file to refer to specific error codes is, as far as I know, undocumented.  I'll defer to @pq and @bwilkerson as to whether or not that feature is considered part of the stable API or not.  I sincerely hope not, since I think it's very bad for the analyzer in general if changing an internal error code is a breaking change, but again, I'll defer to them.  

For the specific case of strong mode, this is definitely an area that we will make changes in.   I basically don't care too much to argue over whether those changes are considered breaking or not.  To provide concrete context: the error code that you were relying on was a catch-all error code for a bunch of errors that we emitted in a separate pass that we tacked on to the analyzer. That catch-all category happened to include mostly things that you wanted to ignore (plus a small number of errors that you didn't really want to ignore and which were hence collateral damage).  That pass is in the process of going away for the sake of code health, performance, and improved user experience.  We are now emitting better, more precise error messages, and we have less duplicated code.  This is an important part of us moving strong mode forward and integrating it into the language.

I understand that this is causing you some pain here, since you were relying on that organization of the errors for your filtering, and I'm willing to work with you to get to a good resolution.  I'm not in general ruling out temporarily reverting changes pending a resolution, but I will only do that very reluctantly - it has a very bad effect on our velocity, and our willingness to make changes in general.  This is particularly true if, as I understand is the case, Flutter is proposing to roll weekly.  

In this particular case, I see a small number of errors, a few of which are things which you will at some point have to address (and can fix for now with casts), and the rest of which are things that we are working on at the language level.  The latter cases are, I believe, in non-API facing code where a small temporary work-around will address the issue pending a resolution.  I would suggest that the simplest and fastest solution would be to modify your code as I've suggested, with a TODO referencing the language changes.  Feel free to file a language bug and assign it to me.  I can commit to prioritizing getting you a prototype solution, but a final solution with syntax depends on buy in from the full language team.

It's also possible that @pq can suggest an alternative path to enabling you guys to just ignore these errors - I know there are a number of paths to doing so in the analyzer currently, but I don't know how all of these play into the Flutter story.

**Comment by Hixie** [comment]
I think there was some miscommunication above. We have no problem churning the analysis file or improving the code by adding annotations or fixing bugs. The only thing we want to avoid is making the code worse, e.g. by losing type annotations we have today.

Do you happen to know what the codes are that replaced the old code so that we can update the analysis options file to ignore the new codes instead, until there's a feature we can use to help the analyzer? 

Specifically, these errors are IMHO not errors and we should either have annotations to help the analyzer understand they're correct code, or we should currently ignore them until such annotations or language features are available:
```
[warning] The element type '(ImmediateMultiDragGestureRecognizer) → ImmediateMultiDragGestureRecognizer' cannot be assigned to the map value type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (examples/flutter_gallery/lib/demo/animation_demo.dart, line 187, col 46)
[warning] The element type '(ImmediateMultiDragGestureRecognizer) → ImmediateMultiDragGestureRecognizer' cannot be assigned to the map value type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (examples/flutter_gallery/lib/demo/animation_demo.dart, line 353, col 46)
[warning] The element type 'PopupMenuDivider' cannot be assigned to the list type 'PopupMenuEntry<LeaveBehindDemoAction>' (examples/flutter_gallery/lib/demo/leave_behind_demo.dart, line 146, col 15)
[warning] The element type 'PopupMenuDivider' cannot be assigned to the list type 'PopupMenuEntry<String>' (examples/flutter_gallery/lib/demo/menu_demo.dart, line 143, col 17)
[warning] A value of type '(PanGestureRecognizer) → PanGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 284, col 40)
[warning] A value of type '(ScaleGestureRecognizer) → ScaleGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 295, col 42)
[warning] A value of type '(HorizontalDragGestureRecognizer) → HorizontalDragGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 269, col 51)
[warning] A value of type '(DoubleTapGestureRecognizer) → DoubleTapGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 236, col 46)
[warning] A value of type '(LongPressGestureRecognizer) → LongPressGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 243, col 46)
[warning] A value of type '(VerticalDragGestureRecognizer) → VerticalDragGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 254, col 49)
[warning] A value of type '(TapGestureRecognizer) → TapGestureRecognizer' cannot be assigned to a variable of type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/gesture_detector.dart, line 226, col 40)
[warning] The element type '(HorizontalDragGestureRecognizer) → HorizontalDragGestureRecognizer' cannot be assigned to the map value type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/scrollable.dart, line 706, col 46)
[warning] The element type '(VerticalDragGestureRecognizer) → VerticalDragGestureRecognizer' cannot be assigned to the map value type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (packages/flutter/lib/src/widgets/scrollable.dart, line 696, col 44)
[warning] The return type 'State<StatefulWidget>' is not a 'T', as defined by the method 'currentState'. (packages/flutter/lib/src/widgets/unique_widget.dart, line 29, col 12)
[warning] The element type '(GetHealth) → Future<Health>' cannot be assigned to the map value type 'CommandHandlerCallback(Command) → Future<Result>' (packages/flutter_driver/lib/src/extension.dart, line 63, col 21)
[warning] The element type '(Tap) → Future<TapResult>' cannot be assigned to the map value type 'CommandHandlerCallback(Command) → Future<Result>' (packages/flutter_driver/lib/src/extension.dart, line 64, col 14)
[warning] The element type '(ScrollIntoView) → Future<ScrollResult>' cannot be assigned to the map value type 'CommandHandlerCallback(Command) → Future<Result>' (packages/flutter_driver/lib/src/extension.dart, line 67, col 25)
[warning] The element type '(Scroll) → Future<ScrollResult>' cannot be assigned to the map value type 'CommandHandlerCallback(Command) → Future<Result>' (packages/flutter_driver/lib/src/extension.dart, line 66, col 17)
[warning] The element type '(GetText) → Future<GetTextResult>' cannot be assigned to the map value type 'CommandHandlerCallback(Command) → Future<Result>' (packages/flutter_driver/lib/src/extension.dart, line 65, col 19)
[warning] The element type '(WaitFor) → Future<WaitForResult>' cannot be assigned to the map value type 'CommandHandlerCallback(Command) → Future<Result>' (packages/flutter_driver/lib/src/extension.dart, line 70, col 18)
[warning] The element type '(SubmitInputText) → Future<SubmitInputTextResult>' cannot be assigned to the map value type 'CommandHandlerCallback(Command) → Future<Result>' (packages/flutter_driver/lib/src/extension.dart, line 69, col 26)
[warning] The element type '(SetInputText) → Future<SetInputTextResult>' cannot be assigned to the map value type 'CommandHandlerCallback(Command) → Future<Result>' (packages/flutter_driver/lib/src/extension.dart, line 68, col 23)
```

For the following error, I'm not really sure I understand what you want us to do:
```
[warning] The return type 'SynchronousFuture<Map>' is not a 'Future<Map<String, List<String>>>', as defined by the method '_manifestParser'. (packages/flutter/lib/src/services/image_resolution.dart, line 147, col 12)
```

In this specific case, we have a JSON-encoded string which we need to convert to a `Map<String, List<String>>`. We want the code to throw as early as possible in checked mode if the JSON doesn't fit this pattern; in release mode behaviour can be undefined (but shouldn't allow arbitrary code execution). We want this to be efficient, so for example we don't want to have to get the `dynamic->dynamic` map back from `JSON.decode` and then copy all the data just to get the type arguments on the instances right.

Today we're just relying on the fact that you can cast from `Map<dynamic, dynamic>` to `Map<String, List<String>>`, though we're then getting the exceptions much later than we'd like.

Any advice you could give us here on how to improve this code so that it's both what we want and what you want would be great.

**Comment by bwilkerson** [comment]
> The use of that file to refer to specific error codes is, as far as I know, undocumented. I'll defer to @pq and @bwilkerson as to whether or not that feature is considered part of the stable API or not. I sincerely hope not, since I think it's very bad for the analyzer in general if changing an internal error code is a breaking change, but again, I'll defer to them.

Shams is, rather unfortunately, in the process of producing documentation on the analysis options file and it includes the ability to enable and disable various error codes. We ought not to be allowing errors (as opposed to warnings, hints and lints) to be disabled, but that *is* the current state of the world.

And yes, although I fought hard against it, we are currently using the names of the error codes because we don't have any stable identifiers for the errors. Florian started working toward stable identifiers at one point, but there has been no progress for a while.

I'd be happy to have a discussion (outside this issue) about how to clean up the mess we're in wrt error codes, ignore comments, and the analysis options file.

**Comment by pq** [comment]
👍  to a higher bandwidth conversation over VC.

**Comment by leafpetersen** [comment]
> improving the code by adding annotations or fixing bugs. The only thing we want to avoid is making the code worse, e.g. by losing type annotations we have today.

I really do appreciate that.  I do think that if you want to roll frequently you will have to accept some temporary regressions.  I understand that it's easy for temporary regressions to slip and turn into permanent regressions, and so I understand why you might want to avoid this if at all possible.  

> Do you happen to know what the codes are that replaced the old code so that we can update the analysis options file to ignore the new codes instead, until there's a feature we can use to help the analyzer?

I believe that you can get the appropriate codes by passing --machine to the analyzer and looking at the third field.  @pq would probably know better.  Unfortunately, I think this is likely to be a very bad solution for you.  The error code for
```
[warning] The element type '(ImmediateMultiDragGestureRecognizer) → ImmediateMultiDragGestureRecognizer' cannot be assigned to the map value type 'GestureRecognizerFactory(GestureRecognizer) → GestureRecognizer' (examples/flutter_gallery/lib/demo/animation_demo.dart, line 187, col 46)
```
is probably MAP_VALUE_TYPE_NOT_ASSIGNABLE .  If you disable this, you will no longer get any errors associated with maps with bad values, e.g.

```dart
Map<int, int> m = {3 : "world"};
```

For the class of errors above associated with covariant callbacks, how would you feel about the following temporary workaround:

```dart
typedef S F<T, S>(T x);

F/*<A, R>*/ covariant/*<A, R>*/( f(x)) => ((/*=A*/ x) => f(x));

void main() {
  Map<int, F<num, num>> m1 = {4 : covariant((int x) { return 3;})};
}
```

Basically this uses a generic top level function to wrap covariant callbacks at the target type.  This eliminates the error in the short term, and marks these locations as places where an eventual language level solution should be applied.  Thoughts?

For this error:

```
[warning] The return type 'SynchronousFuture<Map>' is not a 'Future<Map<String, List<String>>>', as defined by the method '_manifestParser'. (packages/flutter/lib/src/services/image_resolution.dart, line 147, col 12)
```

The best solution is to wrap this map in a delegating map that checks the contents lazily.  The `typed` constructor from here is basically what you would want:

https://www.dartdocs.org/documentation/collection/1.9.1/collection/DelegatingMap-class.html

For this error:

```
[warning] The return type 'State<StatefulWidget>' is not a 'T', as defined by the method 'currentState'. (packages/flutter/lib/src/widgets/unique_widget.dart, line 29, col 12)
```

I don't know what the intent is, so I don't immediately have a suggestion.  Why is this code right?  If I have a `UniqueWidget<State<MySpecialStatefulWidget>>`, why is it reasonable for `.currentState` to give me back a `State<SomeOtherStatefulWidget>`?

**Comment by Hixie** [comment]
> I do think that if you want to roll frequently you will have to accept some temporary regressions.

I'm not sure that we really need to be rolling that frequently. I'm not familiar with the reasons for desiring the current roll, in particular. @pq, can you elaborate on what it is we need from the latest version?


> I understand that it's easy for temporary regressions to slip and turn into permanent regressions, and so I understand why you might want to avoid this if at all possible.

That is one concern, though not a big one. The bigger concern is that our code is intended to be used as sample code by our users. We want people to walk down the "jump to definition" path eagerly. We want our code to be part of the documentation. As such it has to be extremely readable and as free of hacks as possible. We can have hacks for a few days to go through a transition, but we can't have hacks in for weeks at a time.

(This is the same reason we currently don't use dartfmt. Its output doesn't rise to a high enough quality bar.)


> The error code for [...] is probably MAP_VALUE_TYPE_NOT_ASSIGNABLE .  If you disable this, you will no longer get any errors associated with maps with bad values [...]

That's probably ok for now. I'd rather do that than have to have weird incantations that people won't understand when reading the code.


> The best solution is to wrap this map in a delegating map that checks the contents lazily.  The `typed` constructor from here is basically what you would want [...]

I think this particular case points to the general need to take a `Foo<dynamic>` and dynamically convert it to a `Foo<T>` for a given `T`. I can imagine this being a pretty common need. Would it make sense to have some sort of language support for this? Almost like a static cast operator?


> Why is this code right?  If I have a `UniqueWidget<State<MySpecialStatefulWidget>>`, why is it reasonable for `.currentState` to give me back a `State<SomeOtherStatefulWidget>`?  

Because it's guaranteed that it will return the object it created from its `createState` method, which is defined to return a `T`.

Maybe this code would actually already do the right thing if `GlobalKey` was changed to `GlobalKey<T>` on line 28? @pq, do you want to try that?

**Comment by Hixie** [comment]
> the general need to take a `Foo<dynamic>` and dynamically convert it to a `Foo<T>` for a given `T`

For example, consider that you have a `List<num>`, but you "know" that they're all integers, and you need a `List<int>`.

```dart
  List<num> incoming = <num>[1, 2, 3];
  // function style?
  List<int> outgoing = cast<List<int>>(incoming);
  // operator style?
  List<int> outgoing = cast List<int>(incoming);
  List<int> outgoing = cast incoming; // type implied from lhs, lhs explicit type required
```

The idea being that it would keep the same instance, somehow, but go in and change the generic type `T` of that instance somehow, verifying that everything was still type-correct?

**Comment by pq** [comment]
> I'm not sure that we really need to be rolling that frequently. I'm not familiar with the reasons for desiring the current roll, in particular. @pq, can you elaborate on what it is we need from the latest version?

In general, I prefer frequent rolls for the obvious reasons and historically it's payed off as regular integrations have allowed us to be iterative on things like language features, lints and analyses, detect performance regressions (e.g., that time VM background compilation got turned off), and reduce surprise as the platform evolves (e.g., strong-mode changes, pub tool changes, package version constraints become impossible to solve, etc).

In this particular case, besides general goodness an integration will unblock work on Flutter package summaries support which is key to analysis perf goals we've had in our sights for the last Q+.

> Maybe this code would actually already do the right thing if GlobalKey was changed to GlobalKey<T> on line 28? @pq, do you want to try that?

That gets us a `TYPE_ARGUMENT_NOT_MATCHING_BOUNDS` error: 

`'T' does not extend 'State<StatefulWidget>'lib/src/widgets/unique_widget.dart, line 28`

**Comment by Hixie** [comment]
@pq Maybe change the `UniqueWidget` definition to:
```
abstract class UniqueWidget<T extends State<StatefulWidget>> extends StatefulWidget {
```
...?

**Comment by leafpetersen** [comment]
Following up.  Based on offline conversations with @pq and @Hixie, we will look for a different mechanism to allow Flutter to keep filtering the covariant callback related errors.  If this proves too difficult, I will look into how to temporarily re-enable the old behavior for Flutter.  I will follow up with @Hixie on some potential language proposals to support this kind of pattern.  We (@pq and I) will look into setting up some continuous integration testing of strong mode against the flutter repo so that we can provide more runway for flutter when breaking changes land, and to give us earlier visibility into unintentional breakage.

**Comment by pq** [comment]
Sounds good @leafpetersen.  When you can, could you update with a rough time frame?  In particular, I'm wondering if we should wait or whether we should explore temporary local fixes in the interest of getting us integrated.  Thanks!

**Comment by leafpetersen** [comment]
Here's a doc exploring some language level options for handling these kinds of patterns:

https://docs.google.com/a/google.com/document/d/1RqqM-UMp3WGygYZirHqpwW-nZJpbWw9IVf1UJy6kxZc/edit?usp=sharing

**Comment by leafpetersen** [comment]
@pq Do you mean a time frame for having a solution implemented?  It depends somewhat on how long it takes @Hixie and I to arrive at a mutually agreeable solution.  We may do best to find temporary fixes to let you make progress.

**Comment by pq** [comment]
Thanks @leafpetersen.  I was really just trying to arrive at the short to medium term path and whether it made sense to wait for a prototype rather than commit work-arounds.  Thanks for pushing on this!

**Comment by pq** [comment]
I think the short-term plan forward is to add source-level ignores with a reference to a tracking bug --- more details here: https://github.com/flutter/flutter/issues/5771.

**Comment by pq** [comment]
Source is annotated and build is now green.

It ain't pretty obviously but I think we've got a path forward.

@Hixie PTAL.

File: packages/flutter_driver/lib/src/extension.dart [hunk 2]
```diff
@@ -60,14 +60,14 @@ class FlutterDriverExtension {
...
     _commandHandlers.addAll(<String, CommandHandlerCallback>{
-      'get_health': getHealth,
-      'tap': tap,
-      'get_text': getText,
-      'scroll': scroll,
-      'scrollIntoView': scrollIntoView,
-      'setInputText': _setInputText,
-      'submitInputText': _submitInputText,
-      'waitFor': waitFor,
+      'get_health': getHealth, // ignore: map_value_type_not_assignable, #5771
```
Comment (Hixie) [inline packages/flutter_driver/lib/src/extension.dart]: cc @yjbanov

File: packages/flutter/lib/src/widgets/unique_widget.dart [hunk 3]
```diff
@@ -26,6 +26,6 @@ abstract class UniqueWidget<T extends State> extends StatefulWidget {
   /// Might be null if the widget is not currently in the tree.
   T get currentState {
     GlobalKey globalKey = key;
-    return globalKey.currentState;
+    return globalKey.currentState; //ignore: return_of_invalid_type, #5771
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/unique_widget.dart]: nit: missing space after //

**Comment by Hixie** [comment]
Rather than "`#5771`" I would recommend having the whole URL, as in "`https://github.com/flutter/flutter/issues/5771`". It's much harder to follow the former compared to the latter.

**Comment by Hixie** [comment]
LGTM with the suggested changes.
For now.
If we end up with this code for more than a few weeks, though, we will have to revisit this.

File: packages/flutter/lib/src/widgets/unique_widget.dart (hunk 3 above)
Comment (pq) [inline packages/flutter/lib/src/widgets/unique_widget.dart]: Fixed.

**Comment by pq** [comment]
> For now.

👍

**Comment by yjbanov** [comment]
FYI, this PR broke driver tests ([build history](https://flutter-dashboard.appspot.com/build.html), [log](https://flutter-dashboard.appspot.com/api/get-log?ownerKey=ahNzfmZsdXR0ZXItZGFzaGJvYXJkclgLEglDaGVja2xpc3QiOGZsdXR0ZXIvZmx1dHRlci9kMmZkYTY3N2UxYjI5MTk1MzhkNmRmMTBlYmE4NTk4MDNiMTMxNDYyDAsSBFRhc2sYgICAgICAwAgM)).

The issue seems to be that `package:test` is unable to launch the test.

**Comment by pq** [comment]
r'uh r'oh.  Following up via email...

File: packages/flutter_driver/lib/src/extension.dart (hunk 2 above)
Comment (yjbanov) [inline packages/flutter_driver/lib/src/extension.dart]: What's the solution here given that we also use `avoid_as` lint?

Comment (yjbanov) [inline packages/flutter_driver/lib/src/extension.dart]: Actually, nvm, I can declare a variable and do a narrowing implicit cast.

Comment (Hixie) [inline packages/flutter_driver/lib/src/extension.dart]: We may want to revisit the avoid_as lint. I am not sure the reasons for adding it still apply.

**Pull request merged**